firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +35,115 @@
   /** The event listener for the query that raises events asynchronously. */
   private final ExecutorEventListener<ViewSnapshot> asyncEventListener;
 
+  public static class StopListenerSupportFragment extends android.support.v4.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void addCallback(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private List<Runnable> callbacks = new ArrayList();
+  }
+
+  public static class StopListenerFragment extends android.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void addCallback(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private List<Runnable> callbacks = new ArrayList();
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
+    // Android provides lifecycle callbacks (eg onStop()) that custom `Activity`s can extend. But we
+    // can't take advantage of that, since we need to be usable with a generic Activity. So instead,
+    // we create a custom Fragment, and add that Fragment to the given Activity. When the Activity
+    // stops, it will automatically stop the attached Fragments too.
+    //
+    // One difficulty with this approach is that how you get a Fragment and attach it to an Activity
+    // differs based on the type of Activity. If the Activity is actually a FragmentActivity, then
+    // you must use the android.support.v4.app.FragmentManager to do so. Otherwise, you need to use
+    // the deprecated android.app.FragmentManager.
+    //
+    // Possible improvements:
+    // 1) Allow other lifecycle callbacks other than just 'onStop'.
+    // 2) Use LifecycleOwner (which FragmentActivity implements, but Activity does not) to register
+    //    for lifecycle callbacks instead of creating/attaching a Fragment.
+
+    if (activity instanceof FragmentActivity) {
+      FragmentActivity fragmentActivity = (FragmentActivity) activity;
+
+      StopListenerSupportFragment fragment = null;
+      try {
+        fragment =
+            (StopListenerSupportFragment)
+                fragmentActivity
+                    .getSupportFragmentManager()
+                    .findFragmentByTag(SUPPORT_FRAGMENT_TAG);
+      } catch (ClassCastException e) {
```

Comment:
This could be made considerably more concise by adding a cast method:

```java
private static <T> T castFragment(Class<T> fragmentClass, Object fragment) {
  if (fragmentClass.isInstance(fragment)) {
    return fragmentClass.cast(fragment);
  } else {
    throw new IllegalStateException(...);
  }
}
```

(you can also Class.cast and catch the ClassCastException as you've done).

This also makes it possible to include the result of fragment.getClass().getName() in your exception text.

Now instead of separate declarations assignments, etc you just have 

```java
StopListenerSupportFragment fragment = castFragment(
    StopListenerSupportFragment.class,
    fragmentActivity.getSupportFragmentManager()...);
```